src/apiserver/api_service_impl.cc

```diff
@@ -0,0 +1,371 @@
+        auto sp_info = sql_router_->ShowProcedure(db, sp, &status);
+        if (!sp_info) {
+            writer << err.Set(status.msg);
+            return;
+        }
+
+        const auto& schema_impl = dynamic_cast<const ::hybridse::sdk::SchemaImpl&>(sp_info->GetInputSchema());
+        // Hard copy, and RequestRow needs shared schema
+        auto input_schema = std::make_shared<::hybridse::sdk::SchemaImpl>(schema_impl.GetSchema());
+        auto common_column_indices = std::make_shared<fedb::sdk::ColumnIndicesSet>(input_schema);
+        for (int i = 0; i < input_schema->GetColumnCnt(); ++i) {
+            if (input_schema->IsConstant(i)) {
+                common_column_indices->AddCommonColumnIdx(i);
+            }
+        }
+        auto row_batch = std::make_shared<sdk::SQLRequestRowBatch>(input_schema, common_column_indices);
+        std::set<std::string> col_set;
+        for (decltype(rows.Size()) i = 0; i < rows.Size(); ++i) {
+            if (!rows[i].IsArray()) {
+                writer << err.Set("Invalid input data row");
+                return;
+            }
+            auto row = std::make_shared<sdk::SQLRequestRow>(input_schema, col_set);
+
+            if (!Json2SQLRequestRow(rows[i], common_cols_v->value, row)) {
+                writer << err.Set("Translate to request row failed");
+                return;
+            }
+            row->Build();
+            row_batch->AddRow(row);
+        }
+
+        auto rs = sql_router_->CallSQLBatchRequestProcedure(db, sp, row_batch, &status);
+        if (!rs) {
+            writer << err.Set(status.msg);
+            return;
+        }
+
+        ExecSPResp resp;
+
+        // use output schema to encode data
+        auto& output_schema = sp_info->GetOutputSchema();
+        auto output_simple_schema = TransToSimpleSchema(&output_schema);
+
+        if (document.HasMember("need_schema") && document["need_schema"].IsBool() &&
+            document["need_schema"].GetBool()) {
+            resp.data.schema = output_simple_schema;
+        }
+
+        resp.data.rs = rs;
+        writer << resp;
+    });
+}
+
+void APIServiceImpl::RegisterGetSP() {
+    provider_.get("/dbs/:db_name/procedures/:sp_name", [this](const InterfaceProvider::Params& param,
+                                                              const butil::IOBuf& req_body, JsonWriter& writer) {
+        auto err = GeneralError();
+        auto db_it = param.find("db_name");
+        auto sp_it = param.find("sp_name");
+        if (db_it == param.end() || sp_it == param.end()) {
+            writer << err.Set("Invalid path");
+            return;
+        }
+        auto db = db_it->second;
+        auto sp = sp_it->second;
+
+        hybridse::sdk::Status status;
+        auto sp_info = sql_router_->ShowProcedure(db, sp, &status);
+        if (!sp_info) {
+            writer << err.Set(status.msg);
+            return;
+        }
+
+        GetSPResp resp;
+        resp.data.name = sp_info->GetSpName();
```

Comment:
unnecessary data copy

Reply:
You mean the string copy?

Reply:
string and vector

Reply:
If we really need to avoid data copy, it's better to get the result from sdk. We need to encode the sp info here, and it's not a high freq operation, no need to over-optimize. Clear code is better.

Reply:
Why not use ProcedureInfo to pack json directly. Many responses wrapped in different structs is clean code and easy to read?

Reply:
> Why not use ProcedureInfo to pack json directly. Many responses wrapped in different structs is clean code and easy to read?

The point is that we can JsonRead the response directly, so I transformed the schema. pack is easy, but the unpack is too complicated. This pr will only support JsonWriter for execSp & getSp. Maybe redesign in the next time.